### PR TITLE
[WIP] - headless downloader

### PR DESCRIPTION
**Pull request by dirkhh**
Right now this doesn't do a thing, but it gives us a nice target that
has far fewer dependencies and should contain enough parts to download
stuff from a divecomputer and then sync that with cloud storage.

This was inspired by and is intended as a target for @atdotde 's headless work from several years ago that allows you to build an app on a Raspberry Pi or a similar system that could download dives from a divecomputer and push them into the cloud - for example if you have one of the many non-BLE dive computers that aren't supported on iOS (or one of the pure USB, HID, MTP, etc dive computers that aren't supported on Android).

Likely this isn't super useful for Joe diver without someone then creating a sturdy, "travel ready" device that they could use. But I think that a lot of people might like this, if we get it to the point that after initial configuration it is usable via a web browser.

### Describe the pull request:

- [x] New feature

### Pull request long description:

So far all this does is add a new target DownloaderExecutable that builds a headless (and right now, completely useless) version of Subsurface.
It parses a couple of trivial command line args and has a non-functional CLI download function (that isn't even hooked up yet).
This has only been tested on Linux as all the little computer boards that we might target with this are Linux based.

### Mentions:

@atdotde you should be able to push into this branch, so please feel free to view this as a playground

### CODE REVIEW

**Comment by atdotde** [comment]
Thanks for starting this!

**Comment by dirkhh** [comment]
Of course we now need to start building this at least on a couple of targets as part of the CI/CD pipeline...
Otherwise those tests are a bit pointless :-)
I got distracted by something decidedly less fun (cloud server issues, Windows compiler issues), so I didn't have time to continue this morning as I hoped I would... but it should be pretty easy to get from here to a first test download, I think.

**Comment by atdotde** [comment]
What is  your vision how this is supposed to be used (and in what scenario)? We could start with: Install it on a RaspberryPi that is (already, by other means) connected to the internet (via wifi or cable). With a little bit of glue (this is something I could do without much pain at least if aesthetics are not a concern) there could be a web server running  web page to which you could connect with your phone and select a dive computer and port and then it would download dives from that computer and add those to your cloud account. Maybe that would be a reasonable first goal.

**Comment by dirkhh** [comment]
That almost seems more than phase 1.
My rough thinking was
1) allow downloading from a dive computer and uploading to the cloud completely via command line / config files
2) create a microscopic web service that just starts that process, i.e. user turns on their toy, likely tethered to their phone so it works without a keyboard, and then hits a URL on the device to trigger the download from the attached computer - everything configured before using ssh or a keyboard / monitor
3) allow changing the settings (email, passwd, dive computer, device) from that web service
4) offer local web hotspot to invert the connection logic - but I'm not sure that's worth it, realistically

I think (2) is viable without (3).

**Comment by atdotde** [comment]
I am stuck in two places and could use some ideas: When running this from my Mac, it works until dc_device_open() is called from libdivecomputer.c line 1492 from where the program exits without any further output (no crash) and unfortunately I don't know how to produce debugging symbols for libdivecomputer, so I cannot step into this function with the debugger.

On the other hand, when building this on the RaspberryPi, the linker complains:

> [ 98%] Linking CXX executable subsurface-downloader
/usr/bin/ld: core/libsubsurface_corelib.a(cli-downloader.cpp.o): in function `cliDownloader(char const*, char const*, char const*)':
/home/pi/src/subsurface/core/cli-downloader.cpp:10: undefined reference to `DiveImportedModel::DiveImportedModel(QObject*)'
/usr/bin/ld: /home/pi/src/subsurface/core/cli-downloader.cpp:38: undefined reference to `DiveImportedModel::startDownload()'
/usr/bin/ld: /home/pi/src/subsurface/core/cli-downloader.cpp:10: undefined reference to `DiveImportedModel::downloadFinished()'
/usr/bin/ld: /home/pi/src/subsurface/core/cli-downloader.cpp:10: undefined reference to `DiveImportedModel::staticMetaObject'
collect2: error: ld returned 1 exit status

This goes away if I comment out the body of the function on cli-downloader.cpp but of course in that case, there goes the functionality. 

Any ideas? @dirkhh perhaps?

**Comment by dirkhh** [comment]
debugging libdc on macOS is indeed a pain.
The linker issue in the Pi is weird. It looks like the `DiveImportedModel` isn't compiled correctly as that is simply a signal that should be defined... can you check in your compile output of there's something wrong with the way the `diveimportedmodel.cpp` is compiled?

**Comment by dirkhh** [comment]
Ah, the problem was that we had a helper file in the core lib where it didn't belong which messed up our link order.
I think this should fix things for you.

**Comment by atdotde** [comment]
In this state, it's working for me: I can  read dives from my Vyper and sync them with the cloud via an admittedly spartan web interface. Maybe this is a good time for others to look at this? Paging @dirkhh

**Comment by atdotde** [comment]
And yes, the git history is a mess.

**Comment by dirkhh** [comment]
I'm still in Android / mobile hell.
But I'll take a look. Cleaning up git history is one of the things I am good at, so let's see :-)

**Review by dirkhh — COMMENTED** [review]
Definitely some git cleanup - I can easily do that.
Also, some concern about the instructions. Having people install a million packages in order to be able to build the desktop version and then have them not actually run that is a bit, err, excessive. So that could benefit from a bit of streamlining.

**Comment by dirkhh** [comment]
I think the logical thing to do is to add a `-downloader` flag to the `build.sh` script that only builds the pieces that we actually need. I am somewhat hopeful that that would require fewer dependencies. But maybe I'm focusing on the wrong thing here.

**Comment by dirkhh** [comment]
before I push changes - do you want me to do the cleanup and force-push into this branch. Then you could throw your local version away and start from that clean history?

File: core/libdivecomputer.c [hunk 1]
```diff
@@ -40,7 +40,7 @@
 #include <QtGlobal>
 
 char *dumpfile_name;
-char *logfile_name;
+char *logfile_name = "subsurface-downloader.log";
```
[the remark this answers is not in the document]
Comment (atdotde) [reply core/libdivecomputer.c]: yes

File: subsurface-downloader-main.cpp [hunk 2]
```diff
@@ -91,6 +91,7 @@ int main(int argc, char **argv)
 			// download from that dive computer
 			printf("Downloading dives from %s %s (via %s)\n", prefs.dive_computer.vendor, prefs.dive_computer.product, prefs.dive_computer.device);
 			cliDownloader(prefs.dive_computer.vendor, prefs.dive_computer.product, prefs.dive_computer.device);
+			while(1) ;
```
[the remark this answers is not in the document]
Comment (atdotde) [reply subsurface-downloader-main.cpp]: This was intermediate to wait for the download thread. Got replaced further down.

File: cli-downloader.cpp [hunk 3]
```diff
@@ -36,4 +36,5 @@ void cliDownloader(const char *vendor, const char *product, const char *device)
 
 	// before we start, remember where the dive_table ended
```
[the remark this answers is not in the document]
Comment (atdotde) [reply cli-downloader.cpp:37]: I didn't write that. That got copied.

**Comment by atdotde** [comment]
Yes, feel free to clean and force push. Some of the commits are really small (as I used them to move stuff from my laptop to the pi) and could be merged as well. And a merge commit crept in as well.

Regarding the package list: I dumped a list of all packages installed on the pi. But almost all came with the OS image, some are the usual ones needed to build subsurface and then only a few additional ones (with their dependencies) like a perl module, some networking stuff to the the USB tethering going. This list can be shortened significantly. And of course we don't need any of the GUI stuff, so the (terribly long) build time on the Pi could also be reduced by a lot. But I thought that probably the best way to distribute it in the end would be a .img file where everything is installed and configured already, where you only need to set your local wifi an off you go.

But until then, a lot still needs to be done: It's ugly as hell and there is no error handling so far.

**Comment by dirkhh** [comment]
OK, did a force push with a few cleanups and consistent rewrite of the commit messages.
Is this a good point to do a first merge into master, @atdotde ?

**Comment by atdotde** [comment]
Yes, I think this can be merged without doing any harm.

**Review by dirkhh — COMMENTED** [review]
I like the updated approach to missing packages.
Given that this didn't touch any file that compiles, I'll just merge as is

**Pull request merged**